Applications/Spire/Include/Spire/Ui/ArrayTableModel.hpp

```diff
@@ -71,10 +72,25 @@ namespace Spire {
 
       boost::signals2::connection connect_operation_signal(
         const typename OperationSignal::slot_type& slot) const override;
+
+    private:
+      mutable OperationSignal m_operation_signal;
+      std::vector<std::vector<std::any>> m_data;
+      Transaction m_transaction;
+      int m_transaction_level;
   };
 
   template<typename F>
   decltype(auto) ArrayTableModel::transact(F&& transaction) {
+    m_transaction_level++;
+    BOOST_SCOPE_EXIT(this_) {
```

Comment:
The transact function should return the result of transaction, and after calling transaction, something should be done, for example, whether to emit a signal of transaction when all transactions are over (when m_transaction_level == 0).  The code of the transact function is equivalent to the code snippet below:

`m_transaction_level++;
 auto result = transaction();
 m_transaction_level--;
 m_transaction_level == 0) {
        m_operation_signal(this_->m_transaction);
        m_transaction.m_operations.clear();
}
return result;`

The problem of this code snippet is that when the return type of the transaction is void, the statement "return result;" doesn't work because the void type can't be returned. 

I didn't find a way to eliminate the compiler error except to return transaction() directly. In order to execute subsequent code after returning transaction(), I chose BOOST_SCOPE_EXIT which can do something before the end of the transact function.